lib/core/test/unit/Cardano/Wallet/Primitive/MigrationSpec.hs

```diff
@@ -0,0 +1,174 @@
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE NumericUnderscores #-}
+{-# LANGUAGE OverloadedLabels #-}
+
+module Cardano.Wallet.Primitive.MigrationSpec
+    where
+
+import Prelude
+
+import Cardano.Wallet.Primitive.Migration
+    ( RewardWithdrawal (..), createPlan )
+import Cardano.Wallet.Primitive.Migration.Planning
+    ( categorizeUTxO, uncategorizeUTxO )
+import Cardano.Wallet.Primitive.Migration.SelectionSpec
+    ( MockTxConstraints
+    , genCoinRange
+    , genMockTxConstraints
+    , genTokenBundleMixed
```

Comment:
I think this is technically against our coding standards:

> [**Test files are separated and self-contained**](https://github.com/input-output-hk/adrestia/wiki/Coding-Standards#test-files-are-separated-and-self-contained)
> It is really easy to make the testing code more complex than the actual code it's initially testing. Limiting the interaction between test modules helps keeping a good maintainability and a rather low overhead when it comes to extend, modify, read or comprehend some tests. Also, in many cases, we do actually want to have different arbitrary generators for different test cases so sharing instances is risky and cumbersome.

Reply:
> Limiting the interaction between test modules helps keeping a good maintainability and a rather low overhead when it comes to extend, modify, read or comprehend some tests.

I think there's a clear tension between **inter-module interaction** (which we usually want to minimize) and **code reuse** (which might be desirable).

I agree we **definitely** want to avoid the problem of "spooky action at a distance". (Where we change something in one module that has a nefarious effect on another module.)

In the past, we had many `Arbitrary` instances that were shared between test modules.With shared `Arbitrary` instances:
 - When looking at a given call to `arbitrary`, it's often hard to tell where the instance is defined. You first need to know the type, and then need to check the module dependency graph.
 - When changing the definition of an instance, it's hard to tell which other modules and tests will be affected.
 - instances may be defined in terms of more calls to `arbitrary`, which makes solving the instance resolution graph even harder.
  
In this PR, we've not shared any `Arbitrary` instances between modules. (Except, perhaps, for the `Arbitrary (NonEmpty a)` instance, which is relatively innocuous.

In every case of code sharing, we've explicitly imported just the functions we need. So it should be easy for a reader to figure out where something is coming from.

In my mind, this is a good balance: we get the code reuse we want, and the reader can see explicitly where each piece of code comes from.

Having said this, I think we might end up (in the next PR), moving some of these generators to a `Gen` module, so that we can reuse them from the Shelley test suite.